### PR TITLE
Upgrade: [AEA-3724] - upgrade react

### PR DESCRIPTION
**Pull request by anthony-nhs**
## Summary

- Routine Change

### Details

- upgrade react dependencies
- refactor epsat e2e tests to make them more reliable

### CODE REVIEW

File: packages/tool/azure/azure-build-pipeline.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ extends:
       - bash: |
           make prepare-for-epsat-release
           # Get latest released validator tag and set as environment variable in deploy templates
-          VALIDATOR_VERSION=$(curl -s https://api.github.com/repos/NHSDigital/validation-service-fhir-r4/tags | jq -r '.[0].name')
+          VALIDATOR_VERSION=v1.0.165-alpha
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/azure/azure-build-pipeline.yml:35]: its already hard coded where it is used in coordinator. 
This tag is only used to display the version of validator on the footer of epsat - currently it shows 'v-java-upgrade' as the curl command does not get the correct tag

File: scripts/check_language_versions.sh [hunk 2]
```diff
@@ -79,6 +80,11 @@ if [[ "$DEVCONTAINER_VALIDATOR_VERSION" != "$API_BUILD_PIPELINE_VALIDATOR_VERSIO
     FAILED_CHECK=1
 fi
 
+if [[ "$DEVCONTAINER_VALIDATOR_VERSION" != "$EPSAT_VALIDATOR_VERSION" ]]; then
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply scripts/check_language_versions.sh:83]: yes - added

File: packages/tool/site/client/tests/components/claim/claimForm.spec.tsx [hunk 3]
```diff
@@ -109,15 +112,17 @@ async function removeEndorsement() {
 
 async function enterValuesInAllFields() {
   await addEndorsement()
-  userEvent.click(screen.getByLabelText("Patient Paid"))
-  userEvent.selectOptions(screen.getByLabelText("Endorsement 1 Type"), "IP")
-  userEvent.type(screen.getByLabelText("Endorsement 1 Supporting Information"), "£210.91,100ml,Specials Ltd,Lic12345678,BN12345678")
-  userEvent.selectOptions(screen.getByLabelText("Exemption Status"), "0005")
-  userEvent.click(screen.getByLabelText("Evidence Seen"))
+  await React.act(async () => {
+    await userEvent.click(screen.getByLabelText("Patient Paid"))
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/components/claim/claimForm.spec.tsx:116]: did you do a `make deep-clean install` as it does not on mine? 
React.act returns a promise so needs an await
userEvent.click returns a promise so needs an await

File: packages/tool/e2e-tests/helpers.ts [hunk 4]
```diff
@@ -69,6 +68,22 @@ export const fiveTimesDefaultWaitTimeout = defaultWaitTimeout * 5
 export const tenTimesDefaultWaitTimeout = defaultWaitTimeout * 10
 export const apiTimeout = 240000
 
+export async function getElement(
+  driver: ThenableWebDriver,
+  locator: Locator
+): Promise<WebElement> {
+  try {
+    await driver.wait(until.elementLocated(locator), tenTimesDefaultWaitTimeout, `Timeout waiting for ${JSON.stringify(locator)} to be located`)
+      .then(async el => {
```
Comment (MatthewPopat-NHS) [inline packages/tool/e2e-tests/helpers.ts:77]: why are we mixing .then and await

Comment (anthony-nhs) [reply packages/tool/e2e-tests/helpers.ts]: changed

File: packages/tool/site/client/tests/pages/dispensePage.spec.tsx [hunk 5]
```diff
@@ -109,13 +110,16 @@ test("Displays the amend id when amending a dispense notification", async () =>
...
-  renderWithContext(<DispensePage prescriptionId={prescriptionId} amendId="test-id"/>, context)
+  renderWithContext(<MemoryRouter><DispensePage prescriptionId={prescriptionId} amendId="test-id"/></MemoryRouter>, context)
 
   expect(await screen.findByText("Amending Dispense: test-id")).toBeTruthy()
 })
 
 async function renderPage() {
-  const {container} = renderWithContext(<DispensePage prescriptionId={prescriptionId} amendId={null}/>, context)
+  let container
+  await React.act(async () => {
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/pages/dispensePage.spec.tsx]: changed

File: packages/tool/site/client/tests/pages/__snapshots__/sendPage.spec.tsx.snap [hunk 6]
```diff
@@ -1,226 +1,9 @@
...
-        <div class="nhsuk-action-link"><a class="nhsuk-action-link__link" href="baseUrl/view?prescription_id=010E34-A99968-467D9Z"><svg class="nhsuk-icon nhsuk-icon__arrow-right-circle" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" height="36" width="36">
-              <path d="M0 0h24v24H0z" fill="none"></path>
-              <path d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"></path>
-            </svg><span class="nhsuk-action-link__text">View</span></a></div>
-      </td>
-    </tr>
-  </tbody>
-</table>"
-`;
+exports[`Displays confirmation page if multiple prescriptions are sent successfully 1`] = `""`;
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/pages/__snapshots__/sendPage.spec.tsx.snap]: fixed

File: packages/tool/site/client/tests/pages/__snapshots__/doseToTextPage.spec.tsx.snap [hunk 7]
```diff
@@ -37,51 +37,15 @@ exports[`Displays dose to text result 1`] = `
   <details class="nhsuk-details nhsuk-expander">
     <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">1 - FHIR Request that a supplier would send to the FHIR API</span></summary>
     <div class="nhsuk-details__text">
-      <div class="sc-eqUzNf chXQhu"><button class="nhsuk-button sc-iGgVNO cEsNDy" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-iGgVNO cEsNDy" role="button" aria-disabled="false" draggable="false" href="data:application/json;charset=utf-8,%7B%0A%20%20%22req%22%3A%20%22JSON%20Request%22%0A%7D" download="message">Download</a></div>
-      <div class="react-json-view" style="font-family: monospace; cursor: default; background-color: rgb(255, 255, 255); position: relative; margin-top: 10px;">
-        <div class="pretty-json-container object-container">
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/pages/__snapshots__/doseToTextPage.spec.tsx.snap:42]: this is pretty printed now

File: packages/tool/site/client/tests/pages/__snapshots__/dispensePage.spec.tsx.snap [hunk 8]
```diff
@@ -96,274 +96,17 @@ exports[`Displays an error on invalid response 1`] = `
...
-        </div>
-      </div>
-    </div>
-  </div>
-  </div>
-</details>
-</div>
-<div class="sc-eqUzNf chXQhu"><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>"
-`;
+exports[`Displays dispense result 1`] = `""`;
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/pages/__snapshots__/dispensePage.spec.tsx.snap]: fixed

File: packages/tool/site/client/tests/components/dispense/__snapshots__/dispenseForm.spec.tsx.snap [hunk 9]
```diff
@@ -843,110 +843,4 @@ exports[`Reason field is shown when status is set to not dispensed 1`] = `
...
-              <option value="0007">Not Dispensed</option>
-            </select></div>
-        </fieldset>
-      </div>
-    </fieldset>
-  </div>
-  <div class="sc-aYaIB eQquhv"><button class="nhsuk-button" aria-disabled="false" type="submit">Dispense</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
-</form>"
-`;
+exports[`Reason field value is reset when hidden 1`] = `""`;
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/components/dispense/__snapshots__/dispenseForm.spec.tsx.snap]: fixed

File: packages/tool/site/client/tests/components/dispense/dispenseForm.spec.tsx [hunk 10]
```diff
@@ -95,16 +106,23 @@ test("Quantity field is hidden when status is not set to partial dispensed", asy
 })
 
 test("Reason field value is reset when hidden", async () => {
-  const {container} = render(
-    <DispenseForm lineItems={staticLineItemInfoArray} prescription={staticPrescriptionInfo} onSubmit={jest.fn}/>
-  )
+  let container
+  await React.act(async () => {
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/components/dispense/dispenseForm.spec.tsx]: fixed

File: packages/tool/site/client/tests/components/claim/claimForm.spec.tsx [hunk 11]
```diff
@@ -61,11 +62,13 @@ test("Clicking Remove Endorsement button removes one set of endorsement fields",
 
 test("Clicking Claim button calls the callback with form values", async () => {
   const submit = jest.fn()
-  render(<ClaimForm initialValues={noPriorClaimInitialValues} onSubmit={submit}/>)
-
+  await React.act(async () => {
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/site/client/tests/components/claim/claimForm.spec.tsx]: fixed

File: packages/tool/e2e-tests/dispense/returnPrescription.spec.ts [hunk 12]
```diff
@@ -18,10 +19,12 @@ describe("firefox", () => {
 })
 
 async function returnPrescriptionUserJourney(driver: ThenableWebDriver) {
-  await driver.findElement(By.linkText("Return prescription")).click()
+  await (await getElement(driver, By.linkText("Return prescription"))).click()
   await driver.wait(until.elementsLocated(returnPageTitle), defaultWaitTimeout)
-  await driver.wait(until.elementsLocated(returnButton), defaultWaitTimeout)
-  await driver.findElement(returnButton).click()
+  // wait 2 seconds for page to finish rendering
+  await new Promise(r => setTimeout(r, 2000))
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/e2e-tests/dispense/returnPrescription.spec.ts:25]: fixed

File: packages/tool/e2e-tests/validator/validateFhirResource.spec.ts [hunk 13]
```diff
@@ -19,10 +24,19 @@ describe("firefox", () => {
 async function validateFhirResourceUserJourney(
   driver: ThenableWebDriver
 ): Promise<void> {
-  await driver.wait(until.elementLocated(By.linkText("Validate a FHIR Resource")), threeTimesDefaultWaitTimeout)
-  await driver.findElement(By.linkText("Validate a FHIR Resource")).click()
-  await driver.findElement(By.id("validatePayload")).sendKeys(Key.CONTROL, "v")
-  await driver.findElement({xpath: "//*[text() = 'Validate']"}).click()
+  finaliseWebAction(driver, "CLICKING THE LINK FOR Validate a FHIR Resource...")
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/tool/e2e-tests/validator/validateFhirResource.spec.ts:27]: I like having it as a separate function so we can do extra things or customise the logging if needed (eg dont use console.log)

[the remark this answers is not in the document]
Comment (originalphil) [reply packages/tool/e2e-tests/validator/validateFhirResource.spec.ts:28]: Aye, makes sense.

File: packages/tool/site/client/tests/components/claim/claimForm.spec.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (originalphil) [reply packages/tool/site/client/tests/components/claim/claimForm.spec.tsx]: Yeah, just me. Those have gone away after a rebuild.

File: packages/tool/site/client/tests/pages/__snapshots__/doseToTextPage.spec.tsx.snap (hunk 7 above)
[the remark this answers is not in the document]
Comment (originalphil) [reply packages/tool/site/client/tests/pages/__snapshots__/doseToTextPage.spec.tsx.snap:42]: Yay!

**Review by originalphil — APPROVED** [review]

**Pull request merged**